rfc/0001-crashtracker-user-visable-log.md

```diff
@@ -0,0 +1,703 @@
+# RFC 0001: Crashtracker Structured Log Format (v1)
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in [IETF RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119).
+
+## Summary
+This document describes version 1 of the crashinfo data format.
+
+## Motivation
+The `libdatadog` crashtracker detects program crashes.
+It automatically collects information relevant to the characterizing and debugging the crash, including stack-traces, the crash-type (e.g. SIGSIGV, SIGBUS, etc) crash, the library version, etc.
+This RFC establishes a standardized logging format for reporting this information.
+
+### Why structured json
+As a text-based format, json can be written to standard logging endpoints.
+It is (somewhat) human readable, so users can directly interpret the crash info off their log if necessary.
+As a structured format, it avoids the ambiguity of standard semi-structured stacktrace formats (as used by e.g. Java, .Net, etc).
+Due to the use of native extensions, it is possible for a single stack-trace to include frames from multiple languages (e.g. python may call C code, which calls Rust code, etc).
+Having a single structured format allows us to work across languages.
+
+## Proposed format
+A natural language description of the proposed json format is given here.
+An example is given in Appendix A, and the schema is given in Appendix B.
+
+### Required fields
+- `errortype`:
+    Currently, the only value is "crash", but this allows for extension to capture unhandled soft-errors, e.g. "panic", "uncaught exception", etc.
+- `incomplete`:
+    Boolean `false` if the crashreport is complete (i.e. contains all intended data), `true` if there is expected missing data.
+    This can happen becasue the crashtracker is architected to stream data to an out of process receiver, allowing a partial crash report to be emitted even in the case where the crashtracker itself crashed during stack trace collection.
+    This MUST be set to `true` if any required field is missing.
+- `metadata`:
+    Metadata about the system in which the crash occurred:
+    - `library_name`
+    - `library_version`
+    - `family`
+    - `tags`:
+      A set of key:value pairs, representing any tags the crashtracking system wishes to associate with the crash.
+      Examples would include "hostname", "service", and any configuration information the system wishes to track.
+- `os_info`: 
+    The OS + processor architecture on which the crash occurred.
+- `stacktrace`: 
```

Comment:
I think it would make sense to move this under the `error` attribute, `additional_stacktraces` too, and also add:
- `"error.is_crash": true` to indicate that this error is a crash (adds a marker and allows to filter for crashes in the Error Tracking UI)
- `error.kind` and `error.message` for the type and message to show respectively (can be made up, on iOS crashes we base this on the signal; for example `SIGABRT (#0)` for kind and `Application crash: SIGABRT (Abort trap)` for message).
- Perhaps an `"error.stacktrace_format": "crashtracker"` so that we will know the format of the stacktrace and how to parse it. (if we can't determine this from the `source` or another field already)

So it becomes closer to [how we recommend customers](https://docs.datadoghq.com/logs/log_collection/?tab=host#attributes-for-stack-traces) do this.

Adding these attributes can start including events into Error Tracking already, fingerprinted based on these attributes (but maybe that's a happy accident? only the stack trace would not show in the UI; you'd only see the kind and message for now)

@GianlucaBortoli

Reply:
Totally agree w/ Milan.

> Perhaps an "error.stacktrace_format": "crashtracker" so that we will know the format of the stacktrace and how to parse it. (if we can't determine this from the source or another field already)

if it's a "proper" logs integration, we could definitely use `source` for this instead of yet another attribute IMO

Reply:
done